File: src/main/java/com/my/admin/app/commerce/service/SaleService.java
package com.my.admin.app.commerce.service;

import com.black.food.manager.dao.SaleDao;
import com.black.food.manager.model.api.request.client.CreateClientRequest;
import com.black.food.manager.model.api.request.order.CreateOrderRequest;
import com.black.food.manager.model.api.request.sale.CreateSaleRequest;
import com.black.food.manager.model.api.response.SaleResponse;
import com.black.food.manager.model.entity.*;
import com.black.food.manager.model.transformer.SaleTransformer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class SaleService {

    private static final Logger LOGGER = LoggerFactory.getLogger(SaleService.class);

    private SaleDao saleDao;

    private CashService cashService;
    private ClientService clientService;
    private OrderService orderService;
    private PaymentTypeService paymentTypeService;
    private ExceptionService exceptionService;

    private SaleTransformer saleTransformer;

    public Set<SaleResponse> salesFilter(final ZonedDateTime dateFrom, final ZonedDateTime dateTo, final Long userId,
                                         final Integer clientDni, final Long productTypeId, final Long productId, final Long paymentTypeId) {


        var saleFiltered = saleDao.salesFilter(dateFrom, dateTo, userId, clientDni, productId, productTypeId, paymentTypeId);

        saleFiltered.stream()
                .findFirst().orElseThrow(() -> exceptionService.throwNotFoundException("No se encontraron ventas para los filtros seleccionados."));

        return saleFiltered
                .stream()
                .map(SaleTransformer.FROM_SALE_TO_SALE_RESPONSE)
                .collect(Collectors.toSet());
    }

    @Transactional
    public SaleResponse save(final CreateSaleRequest createSaleRequest) {
        Sale sale = new Sale();

        Cash cash = createCash(createSaleRequest.getCashId());
        Client client = createClient(createSaleRequest.getCreateClientRequest());
        Order order = createOrder(createSaleRequest.getCreateOrderRequest());
        PaymentType paymentType = createPaymentType(createSaleRequest.getPaymentTypeId());

        sale.setCash(cash);
        sale.setOrder(order);
        sale.setClient(client);
        sale.setDate(getNowZonedDateTime());
        sale.setPaymentType(paymentType);
        sale.setPayment(createSaleRequest.getPayment());
        sale.setReturned(createSaleRequest.getReturned());
        sale.setTotal(createSaleRequest.getTotal());

        var savedSale = saveSale(sale);

        return saleTransformer.transformSaleToSaleResponse(savedSale);
    }

    @Transactional
    public Sale saveSale(final Sale sale) {
        Sale saleSaved;
        try {
            saleSaved = saleDao.save(sale);
        } catch (Exception e) {
            throw exceptionService.throwConflictException("Ocurrío un error al registrar la venta.");
        }
        return saleSaved;
    }

    public Cash createCash(final Long cashId) {
        return cashService.findById(cashId);
    }

    public Client createClient(final CreateClientRequest createClientRequest) {
        return clientService.findByDniOrElseSave(createClientRequest);
    }

    public Order createOrder(final CreateOrderRequest createOrderRequest) {

        Set<OrderItem> orderItems = createOrderItems(createOrderRequest);
        OrderType orderType = findOrderTypeById(createOrderRequest.getOrderTypeId());

        Order order = new Order();
        order.setOrderItems(new ArrayList<>(orderItems));
        order.setOrderType(orderType);

        return order;
    }

    private Set<OrderItem> createOrderItems(final CreateOrderRequest createOrderRequest) {
        return orderService.createOrderItems(createOrderRequest);
    }

    public PaymentType createPaymentType(final Long paymentTypeId) {
        return paymentTypeService.findById(paymentTypeId);
    }

    private ZonedDateTime getNowZonedDateTime() {
        return ZonedDateTime.now();
    }

    private OrderType findOrderTypeById(final Long orderTypeId) {
        return orderService.findOrderTypeById(orderTypeId);
    }

    @Autowired
    public void setCashService(final CashService cashService) {
        this.cashService = cashService;
    }

    @Autowired
    public void setSaleDao(final SaleDao saleDao) {
        this.saleDao = saleDao;
    }

    @Autowired
    public void setClientService(final ClientService clientService) {
        this.clientService = clientService;
    }

    @Autowired
    public void setOrderService(final OrderService orderService) {
        this.orderService = orderService;
    }

    @Autowired
    public void setPaymentTypeService(final PaymentTypeService paymentTypeService) {
        this.paymentTypeService = paymentTypeService;
    }

    @Autowired
    public void setSaleTransformer(final SaleTransformer saleTransformer) {
        this.saleTransformer = saleTransformer;
    }

    @Autowired
    public void setExceptionService(final ExceptionService exceptionService) {
        this.exceptionService = exceptionService;
    }
}

File: src/main/kotlin/com/bk/reactive/app/commerce/admin/my/service/SaleService.kt
package com.bk.reactive.app.commerce.admin.my.service

import com.bk.reactive.app.commerce.admin.my.dao.SaleDAO
import com.bk.reactive.app.commerce.admin.my.exception.api.FilterSaleException
import com.bk.reactive.app.commerce.admin.my.exception.api.NotFoundSaleException
import com.bk.reactive.app.commerce.admin.my.exception.api.SaleException
import com.bk.reactive.app.commerce.admin.my.exception.api.SaveSaleException
import com.bk.reactive.app.commerce.admin.my.model.api.request.client.CreateClientRequest
import com.bk.reactive.app.commerce.admin.my.model.api.request.order.CreateOrderRequest
import com.bk.reactive.app.commerce.admin.my.model.api.request.sale.CreateSaleRequest
import com.bk.reactive.app.commerce.admin.my.model.api.response.SaleResponse
import com.bk.reactive.app.commerce.admin.my.model.document.Cash
import com.bk.reactive.app.commerce.admin.my.model.document.Client
import com.bk.reactive.app.commerce.admin.my.model.document.Order
import com.bk.reactive.app.commerce.admin.my.model.document.OrderItem
import com.bk.reactive.app.commerce.admin.my.model.document.OrderType
import com.bk.reactive.app.commerce.admin.my.model.document.PaymentType
import com.bk.reactive.app.commerce.admin.my.model.document.Sale
import com.bk.reactive.app.commerce.admin.my.model.mapper.SaleMapper
import mu.KotlinLogging
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.kotlin.core.util.function.component1
import reactor.kotlin.core.util.function.component2
import reactor.kotlin.core.util.function.component3
import reactor.kotlin.core.util.function.component4
import java.time.ZonedDateTime

@Service
class SaleService(
    private val saleDAO: SaleDAO,

    private val cashService: CashService,
    private val clientService: ClientService,
    private val orderService: OrderService,
    private val paymentTypeService: PaymentTypeService,
) {
    private val logger = KotlinLogging.logger { }

    fun salesFilter(
        dateFrom: ZonedDateTime, dateTo: ZonedDateTime, userId: String?,
        clientDni: String?, productTypeId: String?, productId: String?,
        paymentTypeId: String?
    ): Flux<SaleResponse> {
        return saleDAO.salesFilter(dateFrom, dateTo, userId, clientDni, productId, productTypeId, paymentTypeId)
            .switchIfNotFound()
            .map { SaleMapper.fromSaleToSaleResponse(it) }
            .onErrorSaleFilterResume()
    }

    fun saveSale(createSaleRequest: CreateSaleRequest): Mono<SaleResponse> {
        return Mono.zipDelayError(
            createAndSumCash(createSaleRequest.cashId, createSaleRequest.total),
            createClient(createSaleRequest.createClientRequest),
            createOrder(createSaleRequest.createOrderRequest),
            createPaymentType(createSaleRequest.paymentTypeId)
        ).flatMap { (cash, client, order, paymentType) ->
            saveSale(
                Sale(
                    id = null,
                    cash = cash,
                    order = order,
                    client = client,
                    date = getNowZonedDateTime(),
                    paymentType = paymentType,
                    payment = createSaleRequest.payment,
                    returned = createSaleRequest.returned,
                    total = createSaleRequest.total
                )
            )
        }.map {
            SaleMapper.fromSaleToSaleResponse(it)
        }.onErrorSaveSaleResume()
    }

    fun saveSale(sale: Sale): Mono<Sale> {
        return saleDAO.save(sale)
            .doOnError {
                logger.error { "An error occurred while save a sale : $sale" }
            }
    }

    fun createOrder(createOrderRequest: CreateOrderRequest): Mono<Order> {
        return Mono.zip(
            findOrderTypeById(createOrderRequest.orderTypeId),
            createOrderItems(createOrderRequest).collectList()
        ).map { (orderType, orderItems) ->
            Order(
                id = null,
                orderItems = orderItems,
                orderType = orderType
            )
        }
    }

    private fun createAndSumCash(cashId: String, totalSale: Double): Mono<Cash> {
        return cashService.findById(cashId)
            .flatMap { cash ->
                cash.partialBalance += totalSale
                cashService.updatePartialBalance(cash.id!!, cash)
            }
    }

    private fun createClient(createClientRequest: CreateClientRequest): Mono<Client> =
        clientService.findByDniOrElseSave(createClientRequest)

    private fun createOrderItems(createOrderRequest: CreateOrderRequest): Flux<OrderItem> =
        orderService.createOrderItems(createOrderRequest)

    private fun createPaymentType(paymentTypeId: String): Mono<PaymentType> = paymentTypeService.findById(paymentTypeId)

    private fun findOrderTypeById(orderTypeId: String): Mono<OrderType> = orderService.findOrderTypeById(orderTypeId)

    private fun getNowZonedDateTime(): ZonedDateTime = ZonedDateTime.now()

    private fun Mono<SaleResponse>.onErrorSaveSaleResume(): Mono<SaleResponse> =
        this.onErrorResume {
            logger.error { it.printStackTrace() }
            Mono.error {
                SaveSaleException(
                    code = HttpStatus.CONFLICT,
                    message = it.localizedMessage,
                    suppressedSaleException = it.suppressed
                )
            }
        }

    private fun Flux<Sale>.switchIfNotFound(): Flux<Sale> =
        this.switchIfEmpty(
            Flux.error {
                NotFoundSaleException(
                    message = "No sale for selected filters.",
                    code = HttpStatus.NOT_FOUND
                )
            }
        )

    private fun Flux<SaleResponse>.onErrorSaleFilterResume(): Flux<SaleResponse> =
        this.onErrorResume {
            logger.error { it.printStackTrace() }
            if (it is SaleException) {
                Mono.error { it }
            } else {
                Mono.error {
                    FilterSaleException(
                        message = it.localizedMessage,
                        code = HttpStatus.INTERNAL_SERVER_ERROR
                    )
                }
            }
        }
}